TrueCacheAPI.md

```diff
@@ -0,0 +1,236 @@
+So, that means that a concurrent get and invalidate might either result a cached entry or a missing entry.
+
+As a FYI, Infinispan's implementation of Hibernate 2L offers a strict implementation of this interface.
+Using a component called `PutFromLoadValidator`, it tracks the last time an entry or the cache has been invalidated.
+It then compares it with the Hibernate session start time to decide which came first.
+If the transaction that calls `CachedDomainDataAccess.putFromLoad` is older than the invalidation, invalidation wins.
+Otherwise, `CachedDomainDataAccess.putFromLoad` is allowed to store data into the cache.
+Over the years Hibernate users have come to rely on such behaviour, but such maintaining such strictness has a cost.  
+
+
+## Read-Write Cache
+
+A read-write cache supports a wider range of situations compared to a read-only cache:
+
+It handles updates which act on both the cached value and the value at source.
+It also handles removals which act both the cached value and the value at source.
+
+On top of that, the get function can compare the cache valued and the value fetched from the source.
+This is important in a read-write cache because the versions of the data in the cache and the data source might be different.
+This is not the case in a read-only cache. 
+
+```java
+interface ReadWriteCache<K, V> {
+
+  /**
+   * Checks if the cache contains a value associated with the key.
+   * If the value is present, it's returned immediately wrapped in a CompletionStage.
+   * 
+   * If the value is not present, the function is called asynchronously to fetch the value from an unknown source.
+   * In the presence of concurrent updates, a value might have been cached and updated by the time the fetcher returns.
+   * To handle such situations the user can provide a comparator function to help resolve discrepancies:
+   * 
+   * If a value exists in the cache when the fetcher completes, 
+   * the comparator is called with the fetched value as first parameter,
+   * and the cached value as second parameter.
+   * 
+   * If the comparator returns 0 or a negative number,
+   * it means the fetched value is the same or older than the cached value,
+   * in which case the fetched value is not cached and the cached value is returned wrapped in a CompletionStage.
+   * 
+   * If the comparator returns a positive number,
+   * the fetched value is newer than the cached value,
+   * in which case the fetched value is cached and returned wrapped in a CompletionStage.
+   * 
+   * @param key to look up in the cache
+   * @param fetcher called if not value is associated with the key in the cache
+   * @param comparator called if a cached value exists when the fetcher returns 
+   * @return a future containing the cached value or null if not values exist 
+   */
+  CompletionStage<V> get(K key, Function<K, V> fetcher, Comparator<V> comparator);
```

Comment:
From an asynchronous perspective, should `Function<K, V>` be `Function<K, CompletionStage<V>` instead? 

The problem then is how the comparator, you'd probably need a custom comparator that has something like this?

    CompletionStage<Integer> compare(CompletionStage<T> o1, CompletionStage<T> o2)

Reply:
The comparator should still work as is, we would just have to do use `CompletionStage#thenCombine` internally.

Reply:
@wburns Yes.